interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,956 @@
+	PerSnap map[string]*snapDB
+}
+
+// RuleDB stores a mapping from rule ID to rule, and a tree of rule IDs
+// searchable by user, snap, interface, permission, and pattern variant.
+type RuleDB struct {
+	mutex     sync.RWMutex
+	maxIDMmap maxidmmap.MaxIDMmap
+	ids       map[prompting.IDType]int
+	rules     []*Rule
+	perUser   map[uint32]*userDB
+	// notifyRule is a closure which will be called to record a notice when a
+	// rule is added, patched, or removed.
+	notifyRule func(userID uint32, ruleID prompting.IDType, data map[string]string) error
+}
+
+// New creates a new rule database, loads existing rules from the database file,
+// and returns the populated database.
+//
+// The given notifyRule closure may be called before `New()` returns, if a
+// previously-saved rule has expired or if there are conflicts between rules.
+//
+// The given notifyRule closure will be called when a rule is added, modified,
+// expired, or removed. In order to guarantee the order of notices, notifyRule
+// is called with the prompt DB lock held, so it should not block for a
+// substantial amount of time (such as to lock and modify snapd state).
+func New(notifyRule func(userID uint32, ruleID prompting.IDType, data map[string]string) error) (*RuleDB, error) {
+	maxIDFilepath := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "request-rule-max-id")
+	maxIDMmap, err := maxidmmap.OpenMaxIDMmap(maxIDFilepath)
+	if err != nil {
+		return nil, err
+	}
+	rdb := &RuleDB{
+		maxIDMmap:  maxIDMmap,
+		ids:        make(map[prompting.IDType]int),
+		rules:      make([]*Rule, 0),
+		perUser:    make(map[uint32]*userDB),
+		notifyRule: notifyRule,
+	}
+	if err = rdb.load(); err != nil {
+		logger.Noticef("%v; using new empty rule database", err)
+	}
+	return rdb, nil
+}
+
+// load reads the stored rules from the database file and populates the
+// receiving rule database.
+//
+// Returns an error if an existing prompt DB cannot be loaded. An expired rule
+// or rule conflict does not cause an error. If saving the DB is required but
+// the save fails, returns an error.
+//
+// If an error is returned, the rule database is not modified.
+func (rdb *RuleDB) load() error {
+	target := rdb.dbpath()
+	f, err := os.Open(target)
+	if err != nil {
+		return fmt.Errorf("cannot open rules database file: %w", err)
+	}
+	defer f.Close()
+
+	var ruleList []*Rule
+	err = json.NewDecoder(f).Decode(&ruleList)
+	if err != nil {
+		// TODO: store rules separately per-user, so a corrupted rule for one
+		// user can't impact rules for another user.
+		return fmt.Errorf("cannot read stored prompting rules: %w", err)
+	}
+
+	currTime := time.Now()
+	notifyEveryRule := true
+	return rdb.refreshTreeEnforceConsistency(ruleList, currTime, notifyEveryRule)
+}
+
+// save writes the current state of the rule database to the database file.
+//
+// The caller must ensure that the database lock is held.
+func (rdb *RuleDB) save() error {
+	b, err := json.Marshal(rdb.rules)
```

Comment:
One thing I realized when updating unit tests, is that we don't capture the rules in any top level JSON object, but simply serialize a list here. This may be an issue if we ever need to extend the contents with additional top level metadata. I would suggest encapsulating the rules in a simple wrapper.